[찰리] 3단계구현

kimm-charlie opened the pull request:
안녕하세요 cooper, charlie입니다.

step3까지 완료한후 PR요청 보냅니다.



### 기능목록

- 회원 가입 및 회원 목록 조회
  - 회원 가입 기능 
    - (추가)회원가입을 할때 입력폼에 유효한 입력이 입력되지않으면 텍스트로 알림기능 구현
    - (추가)회원가입시 중복된 id를 입력하면 알림기능 구현
  - 회원 목록 조회 기능
  - 회원 프로필 조회 기능
- 글 쓰기 
  - 글쓰기 기능 
    - (추가)제목,내용이 공백이면 db에 저장되지 않는기능 구현 
  - 글 목록 조회 기능
  - 게시글 상세보기 기능
- 회원정보 수정
  - 회원정보 수정 화면
  - 회원정보 수정 기능
    - (추가)회원정보 수정시 기존 비밀번호를 잘못 입력하면 회원정보 수정이 불가능하도록 구현

- was 배포
  - [배포url](http://54.180.31.11:8080/)

### 변경 사항

1. DTO를 domain으로 domain을 DTO 변환하는 로직이 controller에 존재하는것보다 service단에 존재하는게 맞는 것 같아 service단에 변환하는 로직이 존재하게 했습니다.
2. repository에서 error 발생 코드가 중복되는거 같아 error 발생시점을 repository에서 service로 바꿔줬습니다. 
3. 특정 id를 통해 Article 이나 User를 찾을때 null 값을 return하는 경우가 종종 발생해 return 타입을 Optional로 바꿔줬습니다.
4. 에러를 처리할수있는 GlobalExceptionHandler를 추가해줬습니다. @ControllerAdvice를 통해 모든 오류를 한곳에서 처리할수 있도록했습니다. 

### 수정해야할사항

1. JdbcTemplate을 사용하면 "?"를 통해 쿼리에 필요한 파라미터를 입력받기 때문에 가독성이 떨어져 NamedParameterJdbcTemplate을 사용할 생각입니다.
2. mock을 이용해 테스트 코드를 작성할 예정입니다.

### 궁금한점

1. 저번 리뷰때 테스트코드만을 위한 메서드는 삭제해달라 하셨는데, 그럼 해당메서드를 삭제한후 테스트는 어떻게 돌려야하나요..? 일단 지금해당 메서드가 필요한 test 코드는 주석으로 처리해놨습니다.

### URL 

| 기능                  | URL                | Templates                   | HTTP 메서드 |
| --------------------- | ------------------ | --------------------------- | ----------- |
| 회원 가입 기능        | /users/signUp      | templates/user/form.html    | POST        |
| 회원 목록 조회 기능   | /users             | templates/user/list.html    | GET         |
| 회원 프로필 조회 기능 | /users/{id}        | templates/user/profile.html | GET         |
| 글쓰기 기능           | /article/post      | templates/post/form.html    | POST        |
| 게시글 목록 조회 기능 | /                  | templates/index.html        | GET         |
| 게시글 상세보기 기능  | /articles/{id}     | templates/post/show         | GET         |
| 회원정보 수정 화면    | /users/{id}/form   | templates/user/updateForm   | GET         |
| 회원정보 수정 기능    | /users/{id}/update |                             | PUT         |



### 실행결과

<details markdown="1">
<summary>회원가입시 유효하지 않은 정보를 입력했을때</summary>


![스크린샷 2023-04-07 오후 5.01.51](https://raw.githubusercontent.com/CDBchan/Typora-img/main/img/%E1%84%89%E1%85%B3%E1%84%8F%E1%85%B3%E1%84%85%E1%85%B5%E1%86%AB%E1%84%89%E1%85%A3%E1%86%BA%202023-04-07%20%E1%84%8B%E1%85%A9%E1%84%92%E1%85%AE%205.01.51.png)

</details>

<details markdown="1">
<summary>회원가입시 중복된 id를 입력했을때 </summary>


![스크린샷 2023-04-07 오후 4.58.22](https://raw.githubusercontent.com/CDBchan/Typora-img/main/img/%E1%84%89%E1%85%B3%E1%84%8F%E1%85%B3%E1%84%85%E1%85%B5%E1%86%AB%E1%84%89%E1%85%A3%E1%86%BA%202023-04-07%20%E1%84%8B%E1%85%A9%E1%84%92%E1%85%AE%204.58.22.png)

</details>

<details markdown="1">
<summary>회원수정시 비밀번호를 잘못 입력했을때 </summary>


![스크린샷 2023-04-07 오후 4.58.47](https://raw.githubusercontent.com/CDBchan/Typora-img/main/img/%E1%84%89%E1%85%B3%E1%84%8F%E1%85%B3%E1%84%85%E1%85%B5%E1%86%AB%E1%84%89%E1%85%A3%E1%86%BA%202023-04-07%20%E1%84%8B%E1%85%A9%E1%84%92%E1%85%AE%204.58.47.png)

</details>

<details markdown="1">
<summary>존재하지 않는 user의 profile을 보는 시도를 했을때 </summary>


![스크린샷 2023-04-07 오후 5.41.26](https://raw.githubusercontent.com/CDBchan/Typora-img/main/img/%E1%84%89%E1%85%B3%E1%84%8F%E1%85%B3%E1%84%85%E1%85%B5%E1%86%AB%E1%84%89%E1%85%A3%E1%86%BA%202023-04-07%20%E1%84%8B%E1%85%A9%E1%84%92%E1%85%AE%205.41.26.png)


</details>

kimm-charlie committed:
feat: UserSignUpDTO -> UserDTO, UserListDTO추가

User의 List를 반환할때 비밀번호가 없는 DTO를 사용하기위해 UserListDTO클래스를 만들었다.

kimm-charlie committed:
refactor: Userservice에서 DTO -> domain, domain -> DTO로 변환할수있도록 구현

kimm-charlie committed:
refactor: Articleservice에서 DTO->domain, domain -> DTO로 변환하도록 구현

kimm-charlie committed:
refactor: controller,Service 새성자 주입방식으로 변경

kimm-charlie committed:
refactor: PasswordNotFoundException -> InvalidPasswordException

kimm-charlie committed:
refactor: /login_success -> /login-success

kimm-charlie committed:
refactor: test를 위한 resetSequenceForTest 메서드 삭제

kimm-charlie committed:
refactor: Mapping URL user -> users 로 통일

kimm-charlie committed:
refactor: DTO에서 안쓰는 getter setter 삭제

kimm-charlie committed:
refactor:  test 코드에서 org.assertj.Assertion을 사용하도록 변경

kimm-charlie committed:
refactor:  UserService 생성자 주입방식으로 변경

kimm-charlie committed:
docs: h2를 사용하기위한 설정 추가

kimm-charlie committed:
feat: schema.sql에 테이블 생성

application.properties에 추가한 명령어는 서버를 실행 할때마다 자동적으로 schema를 읽을수 있도록  설정했습니다.

kimm-charlie committed:
feat: JDBCArticleRepository클래스를 통해 DB에 Article 저장 기능 구현

kimm-charlie committed:
refactor: db의 table이름 대문자로 통일

kimm-charlie committed:
feat:JDBCUserRepository클래스를 통해 DB에 user저장 기능 구현

kimm-charlie committed:
refactor: Optional 사용 및 UserService에서 예외처리 

JDBC와 Memory Repository에서 예외처리를 위한 중복코드를 제거하기위해 service에서 예외처리를 한다. 이때 repository에서 null값을 return해야 되는 경우가 생겨 Optional을 사용한다.

kimm-charlie committed:
refactor: getUserByUserId -> getUserById 메서드명 변경

kimm-charlie committed:
refactor: Optional 사용 및 ArticleService에서 예외처리

kimm-charlie committed:
refactor: aws 배포전 리팩토링

한글이 깨지는 경우가 생겨 고쳤습니다.
현재 테스트 코드를 위한 static 변수를 reset해주는 메서드를 삭제한 상태라 ArticleServiceTest의 마지막 테스트를 주석처리해줬습니다.

kimm-charlie committed:
feat: 회원가입에대한 유효성 검사 추가

kimm-charlie committed:
refactor: 예외 메세지는 예외처리 클래스안에서 생성

kimm-charlie committed:
feat: js를 통해 회원가입시 입력이 유효한지 알려주는 기능 추가

kimm-charlie committed:
feat: IdDuplicatedException 추가

중복된 id로 회원가입을 하지 못하게 하기위해 추가했다.

kimm-charlie committed:
feat: 입력값에 대한 검증 및 중복 id입력시 회원가입을 제한하는 기능 구현

kimm-charlie committed:
feat: globalException을 통한 중복id 예외 처리

중복된 id가 입력될경우 id입력칸 밑에 "중복된 아이디가 존재합니다" 문구를 띄워줄수 있도록했다.

kimm-charlie committed:
feat: UserNotFoundException 처리

UserNotFoundException발생시 error 페이지로 이동하도록 했습니다.

kimm-charlie committed:
feat: invalidPasswordException 처리

회원정보를 수정할때 기존비밀번호와 다른 비밀번호가 입력되면 기존비밀번호 입력칸 밑에 "비밀번호가 다릅니다" 문구를 띄울수 있도록했습니다.

kimm-charlie committed:
refactor: Table수정

id,nickName은 최대 12글자로 수정했습니다.

kimm-charlie committed:
feat: 회원정보수정창에서 입력이 유효한지 알려주는 기능 구현

kimm-charlie committed:
feat: profileEditDTO에 valid 추가

kimm-charlie committed:
feat: js를 통해 post시 입력이 유효한지 알려주는 기능 추가

kimm-charlie committed:
feat: post시 제목 또는 내용이 비어있으면 포스팅 제한기능 구현

kimm-charlie committed:
feat: 회원가입, 포스팅 날짜 기능 구현

kimm-charlie committed:
refactor: ArticleDTO 수정에 따른 test코드 수정

kimm-charlie committed:
refactor: Date클래스명 변경 및 util패키지로 이동

kimm-charlie committed:
feat: Article 저장 순서 변경

최신글이 위로 올라올수있도록 변경했다.

kimm-charlie committed:
refactor: requestMapping을 사용하여 중복된 url제거

kimm-charlie committed:
docs:README.md 업데이트

kimm-charlie committed:
refactor: push 수정

git이 꼬여서 수정후 다시 push합니다!

kimm-charlie committed:
refactor: was배포를위한 url 수정

kimm-charlie committed:
test: user조회 실패 테스트수정

kimm-charlie committed:
refactor: queryForObject -> query

DB에 데이터가 존재하지 않을때 validate를 하게되면  EmptyResultDataAccessException 발생하는데 이를 없애기 위해  queryForObject대신 query를 사용했습니다.

kimm-charlie committed:
docs: README.md 업데이트

pbg0205 commented on src/main/java/kr/codesqaud/cafe/repository/impl/MemoryArticleRepository.java:
```diff
@@ -1,19 +1,16 @@
 public class MemoryArticleRepository implements ArticleRepository {
-    private Map<Integer,Article> articleRepository;
+    private final Map<Integer,Article> articleRepository;
     private static int sequence = 1;
```
id 를 할당하는 sequence 를 정적 변수대신 인스턴스 필드로 변경하면 테스트를 위한 코드를 제거할 수 있을 것 같습니다. 혹시 다른 의견 있으시면 말씀해주세요!

kimm-charlie replied:
MemoryArticleRepository는  singleton객체라 static으로 선언해줄 필요가 없었던것 같습니다. 
따라서 sequence를 정적변수대신 인스턴스 필드로 변경했습니다. 감사합니다.!

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -1,57 +1,62 @@
     }
 
-    @GetMapping("/user/signUp")
+    @GetMapping("/signUp")
```
URL 일관성을 위해 케밥 케이스(`/sign-up`) 로 변경해주시면 좋을 것 같습니다

kimm-charlie replied:
케밥 케이스라는게 있었군요! 오탄줄 알았습니다. ㅋㅋ
URL은 케밥 케이스로 변경했습니다.!

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -1,57 +1,62 @@
-        userService.addUser(userSignUpDto.toUser());
+
+    @PostMapping("/signUp")
+    public String signUp(@Valid UserDTO userDto, BindingResult bindingResult) {
```
`@Valid` 는 어떤 예외를 반환하나요??

kimm-charlie replied:
@Valid를 통해 UserDTO에 존재하는 필드들에 대한 유효성을 판단한다 생각했습니다. 그 유효성에 대한 판단결과는 BindingResult에 저장되며, 따라서 bindingReslult.hasErrors()가 참이면 /users로 redirect하지 않고 user/form으로 재이동할수 있도록 했습니다.

pbg0205 replied:
MVC 패턴에서는 `BindingResult` 를 이용하는 방법도 있지만 @ControllerAdvice 를 해결하는 방법에 대해 고민해보면 좋을 것 같아 반환하는 예외에 관해 말씀드렸어요~ 반환하는 예외가 무엇인지 찾아보시는 것을 추천합니다 👍

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -1,57 +1,62 @@
-    public String signUp(UserSignUpDTO userSignUpDto) {
-        userService.addUser(userSignUpDto.toUser());
+
+    @PostMapping("/signUp")
```
URL 일관성을 위해 케밥 케이스(/sign-up) 로 변경해주시면 좋을 것 같습니다

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -1,57 +1,62 @@
     }
 
-    @GetMapping("/users")
+    @GetMapping()
```
`()` 제거해주셔도 될 것 같습니다

pbg0205 commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -1,34 +1,56 @@
 
-    public User getUserByUserId(String Id){
-        return memoryUserRepository.findUserById(Id);
+    public List<UserListDTO> getUserList() {
+        return userRepository.findAll().stream()
+                .map(user -> user.toUserListDTO())
+                .collect(Collectors.toUnmodifiableList());
     }
```
User 객체가 DTO 를 의존하고 있네요~ domain 객체가 dto 를 의존하는 방법과 dto 와 domain 방법 중에 dto 를 선택한 이유에 대해서 말씀 부탁드려요 😃

kimm-charlie replied:
#### domain 객체가 dto를 의존하는 방법과 dto객체가 domain을 의존하는 방법이 존재하는데 그중 domain객체가 dto를 의존하는 방법을 선택한 이유를 물어보시는거겠죠??.. 

- domain객체가 dto를 의존하도록 한이유
dto를 사용하는 목적중 하나는 domain에서 필요한 정보만을 선별함으로써 네트워크를 통해 전송해야하는 데이터 량을 줄이는데 있다 생각했습니다. 따라서 위 경우 사용자의 list에 domain에 존재하는 password를 포함할 필요가 없다 생각했기때문에 domain을 UserListDTO로 변환해주는 과정에서 domain이 dto에 의존하는 현상이 발생했습니다.

pbg0205 replied:
도메인 -> DTO 의존 방향일 때, 고려되는 단점이 아래 두가지라고 생각이 드는데 이 부분에 관해 찰리의 의견이 궁금하네요~ 🤔
1. domain 이 dto 를 의존하고 있는 경우에 특정 도메인 외에 다른 데이터가 dto 에 포함되어야 하는 경우
2. 다른 요구사항이 들어와서 작업을 진행했을 때 dto 로 변환하는 로직으로 인해 domain 커지며 역할이 도메인에 역할이 흐릿해지는 단점

pbg0205 commented on src/main/java/kr/codesqaud/cafe/exception/IdDuplicatedException.java:
```diff
@@ -0,0 +1,8 @@
+
+public class IdDuplicatedException extends RuntimeException{
+    public IdDuplicatedException() {
+        super("중복된 아이디가 존재합니다.");
```
매직 스트링 대신에 상수로 선언해주세요 🫠

kimm-charlie replied:
private static final 을통해 에러메세지를 상수로 선언했습니다!

pbg0205 commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -1,34 +1,56 @@
-        memoryUserRepository.save(user);
+    public void addUser(UserDTO userDTO) {
+        validateId(userDTO.getId());
+        userRepository.save(userDTO.toUser());
     }
 
-    public List<User> getUserList(){
-        return memoryUserRepository.findAll();
+    private void validateId(String id) {
+        if (userRepository.findUserById(id).isPresent())
+            throw new IdDuplicatedException();
     }
```
### 1. 특정 ROW 가 존재하는지를 확인하는 방법에 관한 고민

이 부분은 두가지 접근 방법으로 생각할 수 있을 것 같아요. 두 가지 선택지에서 어떤 방법이 더 좋을 것 같은지 한번 확인해주셔도 좋을 것 같습니다. 
1. 별도의 `UserRepository.exist(Long userId)` 를 선언하는 방법
   - SQL: `SELECT  id FROM USER LIMIT 1` 
2. 기존 `findById(Long userId)` 를 선언하는 방법 (SELECT * FROM USER)
   - SQL: `SELECT * FROM USER`

### 2. 2번 방법의 경우 코드 개선

- 2번 방법의 경우에는 stream api 를 통해 이렇게 사용할 수 있는데 이 방법은 어떨까요?? (`검증하고 유저를 등록한다.`가독성의 영역에서는 찰리 방식도 좋아보여서 간결성과 가독성 중에 선택하시면 될 것 같아요!)

```java
    public void addUser(UserDTO userDTO) {
        userRepository.findUserById(userDTO.getId())
                .orElseThrow(IdDuplicatedException::new);
        
        userRepository.save(userDTO.toUser());
    }
```

kimm-charlie replied:
### 1. 특정 ROW 가 존재하는지를 확인하는 방법에 관한 고민
cooper가 말씀해주신데로 특정 id가 db에 존재하는지  확인하기위해 `SELECT *`을 사용하여 모든 컬럼에 대한 정보를 얻는것 보다 저희가 필요로하는 id에 대한 정보를 얻는것이 더 적합할 것 같아 아래와같이 repository에 exist 메서드를 추가했습니다!
``` java
    @Override
    public boolean exist(String id) {
        boolean exist = jdbcTemplate.query("SELECT id FROM \"USER\" WHERE id = ? LIMIT 1 ",(rs,rn) -> rs.getString("id"),id)
                .stream()
                .findFirst()
                .isPresent();
        return exist;
    }
```

### 2. 코드개선
간결성을 위해 합칠까도 생각해봤는데 user를 추가할때 validate기능이 있다는걸 명확하게 해주고싶어 기존 코드 방식을 유지했습니다.
```java
    public void addUser(UserDTO userDTO) {
        validateId(userDTO.getId());
        userRepository.save(userDTO.toUser());
    }

    private void validateId(String id) {
        if(userRepository.exist(id)){
            throw new IdDuplicatedException();
        }
    }
```

pbg0205 commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -1,34 +1,56 @@
+                .orElseThrow(() -> new UserNotFoundException());
+    }
+
+    public void updateUserByUserId(ProfileEditDTO profileEditDto) {
+        UserDTO userDto = getUserById(profileEditDto.getId());
+
+        if (userDto.getPassword().equals(profileEditDto.getOriPassword())) {
+            userRepository.updateUser(profileEditDto.toUser());
+            return;
+        }
+        throw new InvalidPasswordException(profileEditDto.getId());
     }
```
저는 비밀번호 검증하는 부분을 메서드로 분리하면 좋을 것 같다는 생각이 드는데 찰리 의견이 궁금하네요🤔
```java
    public void updateUserByUserId(ProfileEditDTO profileEditDto) {
        UserDTO userDto = getUserById(profileEditDto.getId());

        if (!matchPassword(profileEditDto, userDto)) {
            throw new InvalidPasswordException(profileEditDto.getId());
        }

        userRepository.updateUser(profileEditDto.toUser());
    }

    private boolean matchPassword(ProfileEditDTO profileEditDto, UserDTO userDto) {
        return Objects.equals(userDto.getPassword(), profileEditDto.getOriPassword());
    }
    
```

kimm-charlie replied:
저도 하나의 메서드에서 검증을하고 업데이트를 하는 2가지 기능을 수행하는 것 보다 검증을하는 기능은 따로 분리해 각 메서드가 하나의 기능만을 수행하도록 하는게 맞다고 생각해 다음과 같이 코드를 수정했습니다.
```java
    public void updateUserByUserId(ProfileEditDTO profileEditDto) {
        UserDTO userDto = getUserById(profileEditDto.getId());

        if (!matchPassword(profileEditDto, userDto)){
            throw new InvalidPasswordException(profileEditDto.getId());
        }

        userRepository.updateUser(profileEditDto.toUser());
    }

    private boolean matchPassword(ProfileEditDTO profileEditDto, UserDTO userDto) {
        return userDto.getPassword().equals(profileEditDto.getOriPassword());
    }
```

### objects.equals
cooper가 작성해주신 코드에서
```java
private boolean matchPassword(ProfileEditDTO profileEditDto, UserDTO userDto) {
        return Objects.equals(userDto.getPassword(), profileEditDto.getOriPassword());
    }
```
userDto.getPassword().equals(profileEditDto.getOriPassword()) 대신 위와같이 `Objects.equals`를 사용하셨는데 따로 이유가 있을까요?
제가 찾아본 결과 
`Objects.equals와 .equals의 차이는 Objects.equals는 null 값을 비교할때 false를 return하지만 .equals는 null값을 비교하면 nullpointerException이 발생한다. ` 
정도의 차이가 있는 것 같은데 profileEditDto와 UserDto가 null검증을 이전에 마쳤다면, 굳이 Object.equals()를 쓰지 않아도 되는걸까요?

pbg0205 replied:
ProfileEditDto 가 어디서 null 검증 여부는 작업자는 알지만 다른 작업자의 경우에는 찾아야하는 번거로움이 있어서 `Objects.equals`  를 사용해 다른 고려사항을 제거하기 위한 목적으로 사용하는 편입니다 😃

pbg0205 commented on src/main/java/kr/codesqaud/cafe/exception/IdDuplicatedException.java:
```diff
@@ -0,0 +1,8 @@
+package kr.codesqaud.cafe.exception;
+
+public class IdDuplicatedException extends RuntimeException{
```
Id 는 너무 포괄적인 의미인 것 같아서 조금 더 구체적인 내용이 좋을 것 같습니다 (e.g. `DuplicatedUseCreateException` , `AlreadyUserExstenceException` )

kimm-charlie replied:
감사합니다 `AlreadyUserExstenceException` 이 위 오류에 더 적합한 이름 같아 수정했습니다.!

pbg0205 commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -1,34 +1,56 @@
+    public UserDTO getUserById(String id) {
+        return userRepository.findUserById(id)
+                .map(User::toUserDTO)
+                .orElseThrow(() -> new UserNotFoundException());
```
메서드 참조 방식을 권장해주세요!

- [[Effective Java] 아이템 43 : 람다보다는 메서드 참조를 사용하라](https://velog.io/@semi-cloud/Effective-Java-%EC%95%84%EC%9D%B4%ED%85%9C-43-%EB%9E%8C%EB%8B%A4%EB%B3%B4%EB%8B%A4%EB%8A%94-%EB%A9%94%EC%84%9C%EB%93%9C-%EC%B0%B8%EC%A1%B0%EB%A5%BC-%EC%82%AC%EC%9A%A9%ED%95%98%EB%9D%BC)

pbg0205 commented on src/main/java/kr/codesqaud/cafe/exception/InvalidPasswordException.java:
```diff
@@ -0,0 +1,12 @@
+public class InvalidPasswordException extends RuntimeException{
+    private String id;
+    public InvalidPasswordException(String id) {
+        super("비밀번호가 일치하지 않습니다.");
```
매직 스트링은 상수로 변경해주세요~

pbg0205 commented on src/main/java/kr/codesqaud/cafe/global/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    @ExceptionHandler(IdDuplicatedException.class)
+    public ModelAndView handleIdDuplicatedException(IdDuplicatedException e) {
```
예외 네이밍도 신경써주세요~ 다른 부분도 확인 부탁드려요

kimm-charlie replied:
이때 말씀하시는 네이밍은 `handleIdDuplicatedException` 메서드명 말씀하시는건가요?? 
아니면 `IdDuplicatedException` 이 예외에 대한 말씀을 하신건가요?? 

예외 네이밍은 IdDuplicatedException보다  AlreadyUserExistenceException이 더 명확한것 같아 변경했습니다.!

pbg0205 replied:
AlreadyUserExistenceException 의 변수명이 e 로 되어 있어요!

pbg0205 commented on src/main/java/kr/codesqaud/cafe/global/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,39 @@
+    @ExceptionHandler(InvalidPasswordException.class)
+    public ModelAndView handleInvalidPasswordException(InvalidPasswordException e) {
+        ModelAndView mav = createErrorResponseModelAndView("user/updateForm", e);
+        mav.addObject("id", e.getId());
```
예외에 userId 를 담아서 전달하는 의도가 궁금합니다 🤔

kimm-charlie replied:
### 예외에 id가 존재하는 이유
회원정보를 수정할때 기존 비밀번호 + 바꿀 비밀번호 그외 변경할 데이터를 입력합니다. 
이때 기존 비밀번호를 잘못입력하게 되면 InvalidPasswordException이 발생하는데
이때 아래와같이 사용자 아이디 칸에 id가 (자동적으로)들어가기 위해선 mustache를 사용하여 updateForm에 id를 동적으로 추가해줘야해 예외에 userId를 추가한후 Model에 id를 담아 view에 전달해줬습니다.
<img width="523" alt="스크린샷 2023-04-07 오후 4 58 47" src="https://user-images.githubusercontent.com/98310007/230731843-3656cebc-7263-4e43-975b-8c26c7a1cd5d.png">

저도 이렇게 만들면서 뭔가 다른 해결 방법이 있을 것 같은데 좀더 다양한 시도를 해봐야 할 것 같습니다!

pbg0205 commented on src/main/java/kr/codesqaud/cafe/exception/InvalidPasswordException.java:
```diff
@@ -0,0 +1,12 @@
+package kr.codesqaud.cafe.exception;
+
+public class InvalidPasswordException extends RuntimeException{
+    private String id;
```
예외 클래스에 userId 필드를 가지고 있는 의도가 궁금합니다 x2

pbg0205 commented on src/main/java/kr/codesqaud/cafe/exception/ArticleNotFoundException.java:
```diff
@@ -1,7 +1,7 @@
-    public ArticleNotFoundException(String message) {
-        super(message);
+    public ArticleNotFoundException() {
+        super("해당 글이 존재하지 않습니다.");
```
상수 선언해주세요!

pbg0205 commented on src/main/java/kr/codesqaud/cafe/global/ErrorResponse.java:
```diff
@@ -0,0 +1,12 @@
+package kr.codesqaud.cafe.global;
+
+public class ErrorResponse {
+    private int status;
+    private String message;
+
```
`final` 을 선언해주면 좋을 것 같습니다~ final 키워드의 장점은 무엇일까요??

kimm-charlie replied:
제가 알고있기론 final을 사용하면 상수 취급을 하게되어 생성자를 제외한 다른 곳에서 해당 변수에 대한 값을 못바꾸는 기능을 하는것으로 알고있습니다. 
사실 final을 사용하면서도 특별한 장점이 뭔지 느끼지 못했는데 찾아보니 변경될 가능성이 있는 필드 외에 final로 선언을 하면 나중에 디버깅시 추적하는게 편하다는데 흠.. 아직 final을 사용해서 느껴본 큰 장점이 없는것 같습니다.

pbg0205 replied:
final 을 사용하면 특정 객체를 데이터를 할당해야 하는 시점을 생성자로 특정할 수 있습니다. 만약에 비록 setter 를 닫아놓고 생성자를 통해서 생성자로 특정할 수 있지만 필드가 추가되는 요구사항에서 값을 누락했을 때는 컴파일 시점에 잡아주지 않습니다~ 직접 찾아야 하는 단점이 있는데 final 을 사용하면 이런 휴먼 에러를 컴파일 시점에서 확인이 가능하고 로직을 응집화할 수 있는 장점이 있어요~

pbg0205 commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -1,34 +1,56 @@
 
-    public UserService() {
-        this.memoryUserRepository = new MemoryUserRepository();
+    public UserService(UserRepository userRepository) {
+        this.userRepository = userRepository;
     }
```
`MemoryUserRepository` 에서 `JdbcUserRepository` 로 변경하시면서 빈 충돌이 있어서 `MemoryUserRepository` 의 @Repository 어노테이션을 제거하신 것 같은데 @Qualifer 를 사용하시면 특정 빈의 의존성을 설정할 수 있습니다! 이와 관련해서 작업해보시고 잘 안돼면 말씀해주세요😃

kimm-charlie replied:
오 좋은정보 감사합니다! 저도 @Repository라는것을 입력하고 삭제하는게 번거로워 찾아보다 @Primary라는 annotation으로 @Repository가 여러개 존재해도 우선순위를 지정할수 있다는것을 알게 됬는데 cooper덕에 @Qualifier라는것도 배워갑니다!
@Primary를 사용하더라도  repository에 들어가 삭제하고 추가하고 하는 작업이 번거로웠는데 @Qualifier를 통해 Service에서 어떠한 걸 사용할지 선택해줄수있는 방법이 더 편한것 같네요!

pbg0205 commented on src/main/java/kr/codesqaud/cafe/util/DateUtil.java:
```diff
@@ -0,0 +1,23 @@
+
+public class DateUtil {
+
+    private String date;
```
유틸 클래스에는 인스턴스 필드를 사용하지 않습니다. 다른 방법으로 날짜를 선언하는 방법에 대해서 고민해주세요~

kimm-charlie replied:
util에 대해 찾아보니 대게 필드를 갖지 않으며, static을 통해 메서드를 구성하고 메서드를 통해 값을 가져온다해 아래와같이 코드를 수정했습니다.
```java
public class DateUtil {

    public static String getCurrentDate() {
        LocalDate date = LocalDate.now();
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
        return date.format(formatter);
    }
}
```

pbg0205 replied:
추가적으로 유틸클래스가 인스턴스화해야 하는지에 관한 부분도 고민해보시면 좋을 것 같습니다. 참고하실만한 내용 첨부드릴게요!
- https://jjingho.tistory.com/46

pbg0205 commented on src/main/java/kr/codesqaud/cafe/domain/User.java:
```diff
@@ -1,17 +1,43 @@
+        this.date = setDate();
+    }
+
+    public User(ResultSet rs) throws SQLException {
+        this.nickName = rs.getString("nickName");
+        this.email = rs.getString("email");
+        this.password = rs.getString("password");
+        this.id = rs.getString("id");
+        this.date = rs.getString("date");
+    }
```
ResultSet 을 파라미터로 전달하는 방법보다 외부에서 받아서 사용하시는 방법은 어떨까요? 도메인 코드가 DB에 의존적이기 때문에 추후에 유지보수 하기가 어려워요😢

kimm-charlie replied:
### "ResultSet 을 파라미터로 전달하는 방법보다 외부에서 받아서 사용하시는 방법은 어떨까요?" 
현재 domain에 DB에 관한 로직이 섞여있어 분리하라는 말씀이신가요?? 찾아보니 DB에 관한 로직은 DAO에서 처리를 해주는 것 같은데 이기회에 DAO에대해 조금더 공부한뒤 적용을해보겠습니다!

pbg0205 commented on src/main/java/kr/codesqaud/cafe/domain/User.java:
```diff
@@ -1,17 +1,43 @@
         this.email = email;
         this.password = password;
         this.id = id;
+        this.date = setDate();
```
만약에 Util 을 사용하신다면 생성자에서 외부로 받는 것이 좋을 것 같습니다!

kimm-charlie replied:
util 클래스의 메서드를 static으로 변경한뒤 아래와같이 생성자에서 외부로 받게 수정했습니다!

```java
public User(String nickName, String email, String password,String id) {
        this.nickName = nickName;
        this.email = email;
        this.password = password;
        this.id = id;
        this.date = getCurrentDate();
    }
```

kimm-charlie replied:
아래 올려주신 util관련 블로그 글을 읽고 DateUtil클래스를 아래와 같이 바꿧습니다.
```java
public class DateUtil {

    private DateUtil(){
    }
    
    public static final String getCurrentDate() {
        LocalDate date = LocalDate.now();
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
        return date.format(formatter);
    }
}
```
private 생성자를 통해 객체생성을 막은뒤 static final 키워드를 사용해 외부에서 메서드를 통해 원하는 값을 가져갈수있도록 만들었습니다!

pbg0205 commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,19 @@
+DROP TABLE IF EXISTS "USER";
+DROP TABLE IF EXISTS ARTICLE;
+
+CREATE TABLE "USER"
```
"USER" 로 선언하신 이유가 궁금합니다🤔

kimm-charlie replied:
h2 DB에서는 USER가 예약어라 테이블명으로 사용하기 위해선 큰따옴표("")로 지정해 위와같이 "USER"로 지정했습니다!

CDBchan replied:
현재 사용하고있는 h2 db에선 USER가 예약어라 테이블명으로 사용하기 위해 "USER"로 선언을 했습니다!

pbg0205 commented on src/main/java/kr/codesqaud/cafe/util/DateUtil.java:
```diff
@@ -0,0 +1,23 @@
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+public class DateUtil {
```
DateUtil 을 사용할 경우에는 final 키워드와 인스턴스화를 막아야 합니다. 관련 자료 참고해보시고 학습해보시면 좋을 것 같습니다 😃

- [[이펙티브 자바] 규칙4. Util 클래스와 같이 객체 생성이 필요없는 클래스는 private 생성자를 사용](https://camel-context.tistory.com/37)

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/dto/UserListDTO.java:
```diff
@@ -0,0 +1,14 @@
+package kr.codesqaud.cafe.controller.dto;
+
+public class UserListDTO {
+    private String nickName;
+    private String email;
+    private String id;
+    private String date;
```
DTO 는 불변을 위해 final 키워드를 추가하시는 것을 추천합니다!

kimm-charlie replied:
DTO 및 domain에 final 키워드를 추가했습니다!

pbg0205 replied:
도메인은 update 로직이 있어서 상태가 변경될 수 있기 때문에 final 은 제거해주셔도 될 것 같아요!

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -17,19 +22,23 @@ public ArticleController(ArticleService articleService) {
 
-    @GetMapping("/article/post")
+    @GetMapping("/post")
     public String showPostArticleForm(){
-        return "/post/form";
+        return "post/form";
     }
```
입력창은 addViewController 에 추가하셨던 걸로 기억하는데 만약에 맞다면 로직 이동하시면 좋을 것 같아요~

kimm-charlie replied:
저번에 배운데로 /post에대한 @PostMapping이 존재할때 
```java
@GetMapping("/post")
    public String showPostArticleForm(){
        return "post/form";
    }
```
이 로직을 addViewController로 옮기면 에러가 발생해 위와같이 Controller안에 showPostArticleForm()메서드를 만들었습니다!

kimm-charlie replied:
cooper가 저번에 알려주신 글 그대로 남깁니다!

같은 URL 일 경우에는 HTTP Method 와 관계없이 우선 순위가 @RequestMapping 을 먼저 검색하기 때문에 검색하지 못합니다. 만약 addViewControllers 로 분리해서 사용하고자 하는 경우에는 Controller 의 URL 과 다르게 선언해야 할 것 같아요

ViewControllerRegistry.java
![](https://user-images.githubusercontent.com/48561660/229357673-82e79add-fcae-4f96-8205-7a8eedab4f54.png)

pbg0205 replied:
화면을 조회하기 위한 목적으로 addViewController 구분하신 걸로 받아들였는데
작성하신 의도를 더욱 분명하게 하기 위해서 addViewController 로 옮기는 것이 좋을 것 같아요~

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -17,19 +22,23 @@ public ArticleController(ArticleService articleService) {
-    public String postArticle(ArticleDTO articleDto) {
-        articleService.post(articleDto.toArticle());
+
+    @PostMapping("/post")
+    public String postArticle(@Valid ArticleDTO articleDto, BindingResult bindingResult) {
```
POST http method 에 생성이라는 의미를 내포하고 있어서 /post 는 제거하셔도 될 것 같습니다~

kimm-charlie replied:
```java
    @GetMapping
    public String showPostArticleForm(){
        return "post/form";
    }

    @PostMapping
    public String postArticle(@Valid ArticleDTO articleDto, BindingResult bindingResult) {
        if(bindingResult.hasErrors()){
            return "post/form";
        }
        articleService.post(articleDto);
        return "redirect:/";
    }
```
위와 같이 바꿧습니다!

pbg0205 reviewed (changes requested):
안녕하세요 찰리~ 리뷰어 cooper 입니다! 일주일동안 많은 작업을 하시느라 수고 많으셨어요ㅎㅎ
전반적으로 컴포넌트 구조를 잘 잡으셔서 쉽게 코드 파악할 수 있었어요. 예외처리까지 작업해주신 부분 확인하면서 이번 미션에서는 예외 처리에 관해 많은 고민해주셨던 것 같아요. 좋은 부분들에 관한 고민을 점점 많이 해주시는 것 같습니다 👍
한가지 당부드리고 싶은 테스트를 위한 코드 제거에 관해 문의 주셨는데 `어떻게 하나요?` 보다는 제거하기 위한 시도나 해결이 되지 않는 이유에 관해 같이 말씀해주시면 좋을 것 같아요. 어떤 노력을 하셨는지 공유해주시면 같이 더 좋은 방법이나 해결방안에 관해 이야기해볼 수 있을 것 같습니다 😃
아래 코멘트에 관해 확인해보시고 궁금한 점이나 같이 고민해보고 싶은 내용 있으시면 언제든 DM 주세요! 화이팅입니다 🫡

kimm-charlie committed:
refactor(리뷰반영): 정적변수 -> 인스턴스 필드로 변경

kimm-charlie committed:
refactor(리뷰반영): 케밥 케이스 적용

kimm-charlie committed:
refactor(리뷰반영): 에러메세지 상수선언

kimm-charlie committed:
refactor(리뷰반영): repository에 id중복 체크를 위한 exist메서드 추가

kimm-charlie committed:
refactor(리뷰반영): updateUserById 메서드 분리

kimm-charlie committed:
refactor(리뷰반영): IdDuplicatedException -> AlreadyUserExistenceException

kimm-charlie committed:
refactor(리뷰반영): 메서드 참조방식으로 수정

kimm-charlie committed:
refactor(리뷰반영): ErrorResponse클래스의 필드 final로 선언

kimm-charlie committed:
refactor(리뷰반영): qualifier 추가

kimm-charlie committed:
refactor(리뷰반영): DateUtil 클래스 수정

kimm-charlie committed:
refactor(리뷰반영): DateUtil 클래스 수정

private 생성자를 통해 외부에서 객체를 만들수 없게 만들었다. 그후 static final 키워드를 사용해 메서드를 통해 원하는 값을 얻을수 있도록 했다.

kimm-charlie committed:
refactor(리뷰반영): 수정하지 않는 필드에 final키워드 추가

kimm-charlie commented:
cooper가 남겨주신 코멘트에 답글남겼습니다! 디테일한 부분까지 짚어주셔서 감사합니다!!😁 덕분에 생각하지 못한 부분까지 더 공부할수있었습니다.
아직 **"ResultSet 을 파라미터로 전달하는 방법보다 외부에서 받아서 사용하시는 방법은 어떨까요?"** 이부분에 대해서는 추가적인 공부가 필요할 것 같아 공부후 수정 하겠습니다! 요번주에도 감사합니다 cooper!!

kimm-charlie committed:
refactor(리뷰반영): post 중복의미 제거

pbg0205 reviewed (approved):
수고 많으셨습니다 찰리! 때로는 조금 깐깐하게 전달드리는 부분도 있는데 항상 긍정적으로 받아들여 주셔서 더욱 적극적으로 리뷰할 수 있는 것 같아요ㅎㅎ 추가적으로 코멘트 드린 부분은 다음 미션에 반영해주시면 될 것 같습니다! 일주일동안 많은 지식 습득하시느라고 피곤하실텐데 푹쉬시고 다음주에 뵐게요 🖐
